firedrake/checkpointing.py

```diff
@@ -824,14 +903,39 @@ def save_function(self, f, idx=None, name=None):
                 _name = "_".join([PREFIX_EMBEDDED, f.name()])
                 _f = Function(_V, name=_name)
                 self._project_function_for_checkpointing(_f, f, method)
-                self.save_function(_f, idx=idx)
+                self.save_function(_f, idx=idx, timestepping_info=timestepping_info)
                 self.set_attr(path, PREFIX_EMBEDDED + "_function", _name)
             else:
                 # -- Save function topology --
                 path = self._path_to_function(tmesh.name, mesh.name, V_name, f.name())
                 self.require_group(path)
                 self.set_attr(path, PREFIX + "_vec", tf.name())
                 self._save_function_topology(tf, idx=idx)
+                # store timstepping_info only if in timestepping mode
+                if idx is not None:
+                    # We make sure the provided timestepping_info is consistent all along timestepping
+                    # For this we store an attribute with "_history_keys" in the first instance
+                    # and we make sure the provided info at each timestep is consistent with that
+                    if self.has_attr(path, PREFIX_TIMESTEPPING + "_history" + "_keys"):
+                        key_names = self.get_attr(path, PREFIX_TIMESTEPPING + "_history" + "_keys")
+                        if key_names != "_".join(timestepping_info.keys()):
+                            raise RuntimeError(
+                                r"Provided keys in timestepping_info must remain consistent")
+                    else:
+                        key_names = "_".join(timestepping_info.keys())
+                        self.set_attr(path, PREFIX_TIMESTEPPING + "_history" + "_keys", key_names)
+
```

Comment:
```suggestion
                if idx is not None:
                    # We make sure the provided timestepping_info is consistent all along timestepping.
                    if not new:
                        existing_keys = {key.replace(PREFIX_TIMESTEPPING_HISTORY + "_", "", 1)
                                         for key in self.h5pyfile[path].attrs.keys()
                                         if key.startswith(PREFIX_TIMESTEPPING_HISTORY)}
                        if timestepping_info.keys() != existing_keys:
                            raise RuntimeError(
                                r"Provided keys in timestepping_info must remain consistent")
```
We can avoid storing a new attribute this way.